### PR TITLE
Add versions used to compile/build against to jar manifest

### PR DESCRIPTION
**Pull request by modmuss50**
TODO tests and cleanup.

### CODE REVIEW

File: src/main/java/net/fabricmc/loom/configuration/JarManifestConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.java.archives.Attributes;
+import org.gradle.jvm.tasks.Jar;
+import org.gradle.util.GradleVersion;
+
+import net.fabricmc.loom.LoomGradleExtension;
+import net.fabricmc.loom.LoomGradlePlugin;
+import net.fabricmc.loom.util.Constants;
+
+public final record JarManifestConfiguration(Project project, Jar jar) {
```
Comment (liach) [inline src/main/java/net/fabricmc/loom/configuration/JarManifestConfiguration.java]:
```suggestion
public record JarManifestConfiguration(Project project, Jar jar) {
```
records are automatically final.

**Comment by Player3324** [comment]
the mapping namespace and maybe associated version could be interesting as well

File: src/main/java/net/fabricmc/loom/configuration/JarManifestConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	public void configure() {
+		// Done in doFirst to ensure that everyone else has had time to modify the manifest
+		// and to ensure that the compile classpath has already been resolved.
+		jar.doFirst(configureManifestAction());
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/JarManifestConfiguration.java]: Doesn't doing this lazily in `doFirst` completely prevent users from customising this? I don't think you should force all jars from the `jar` task to have these attributes necessarily.

Comment (modmuss50) [reply src/main/java/net/fabricmc/loom/configuration/JarManifestConfiguration.java]: I wasnt too sure about this, might move it till after remapping. Will be needed anyway if we are going to include the namespace.

File: src/main/java/net/fabricmc/loom/configuration/JarManifestConfiguration.java (hunk 1 above)
Comment (modmuss50) [reply src/main/java/net/fabricmc/loom/configuration/JarManifestConfiguration.java]: I will see if I can add a checkstyle rule for this. https://checkstyle.sourceforge.io/config_modifier.html#RedundantModifier might be a good idea to have.

**Comment by modmuss50** [comment]
I have now moved this to after remapping.

```
Manifest-Version: 1.0
Fabric-Loom-Version: 0.0.0+unknown
Fabric-Gradle-Version: 7.0.1
Fabric-Mixin-Compile-Extensions-Version: 0.4.4
Fabric-Tiny-Remapper-Version: 0.3.2
Fabric-Loader-Version: 0.11.2
Fabric-Mixin-Version: 0.9.2+mixin.0.8.2
Fabric-Mapping-Namespace: intermediary
```

Is an example of what will be added to the manifest.

**Comment by modmuss50** [comment]
This will mean we can no longer be certain that builds are reproduable across loom patch versions, I think its fine to say if you want that you should depend on a specific loom build 👍

**Review by Player3324 — APPROVED** [review]
LGTM, thanks for tackling this!

**Review by Player3324 — CHANGES_REQUESTED** [review]
Actually I need a way to know whether it was fabric mixin or not (maven group?)

**Comment by Player3324** [comment]
Likewise the version of the mappings, kind of plays into layered mapping territory?

**Comment by modmuss50** [comment]
Sure I can add the mixin group, the actual minecraft version would be useful as well.

The mapping version might be less useful and would just be a hash when using layered mappings. I can add it but it doesn’t seem terribly useful.

**Review by Player3324 — APPROVED** [review]

**Pull request merged**